File: ports/r/conformance_test.R
# conformance_test.R - Check the R port matches Python exactly.
#
# Run:  Rscript ports/r/conformance_test.R
#
# jsonlite is the only dependency and it is only needed to read the fixture;
# the port itself has none.

suppressWarnings(suppressMessages({
  if (!requireNamespace("jsonlite", quietly = TRUE)) {
    cat("R: SKIP - jsonlite is not installed (install.packages('jsonlite'))\n")
    quit(status = 0)
  }
  library(jsonlite)
}))

here <- dirname(sub("^--file=", "", grep("^--file=", commandArgs(), value = TRUE)[1]))
if (is.na(here) || !nzchar(here)) here <- "ports/r"
source(file.path(here, "fingerradius.R"))

fixture_path <- file.path(here, "..", "fixtures", "conformance.json")
fixture <- fromJSON(fixture_path, simplifyVector = FALSE)

TOL <- 1e-6
failures <- character(0)
checks <- 0L

for (item in fixture$gestures) {
  landmarks <- matrix(unlist(item$landmarks), ncol = 2, byrow = TRUE)
  expected <- item$expected
  actual <- fr_classify(landmarks)
  checks <- checks + 1L

  if (!identical(actual$gesture, expected$gesture)) {
    failures <- c(failures, sprintf("%s: gesture %s != %s",
                                    item$id, actual$gesture, expected$gesture))
  }
  if (actual$extended_count != expected$extended_count) {
    failures <- c(failures, sprintf("%s: extended_count %d != %d",
                                    item$id, actual$extended_count,
                                    expected$extended_count))
  }
  for (finger in names(expected$fingers)) {
    want <- isTRUE(expected$fingers[[finger]])
    got <- isTRUE(actual$fingers[[finger]])
    if (got != want) {
      failures <- c(failures, sprintf("%s: finger %s %s != %s",
                                      item$id, finger, got, want))
    }
  }

  radii <- fr_radii(landmarks)
  for (pair in names(expected$radii)) {
    checks <- checks + 1L
    delta <- abs(radii[[pair]] - expected$radii[[pair]])
    if (delta > TOL) {
      failures <- c(failures, sprintf("%s: radius %s delta %.3e",
                                      item$id, pair, delta))
    }
  }
}

kal <- fixture$kalman
filter <- fr_kalman(process_noise = kal$process_noise,
                    measurement_noise = kal$measurement_noise)
worst <- 0
for (i in seq_along(kal$input)) {
  got <- filter$update(kal$input[[i]])
  want <- kal$expected[[i]]
  checks <- checks + 1L
  delta <- abs(got - want)
  worst <- max(worst, delta)
  if (delta > TOL) {
    failures <- c(failures, sprintf("kalman[%d]: delta %.3e", i, delta))
  }
}

if (length(failures) == 0L) {
  cat(sprintf("r: PASS - %d checks against the Python fixture (Kalman worst %.3e)\n",
              checks, worst))
  quit(status = 0)
} else {
  cat(sprintf("r: FAIL - %d of %d checks\n", length(failures), checks))
  for (f in head(failures, 10)) cat("  ", f, "\n")
  quit(status = 1)
}

File: ports/ruby/conformance_test.rb
# frozen_string_literal: true

# conformance_test.rb - Check the Ruby port matches Python exactly.
#
# Run:  ruby ports/ruby/conformance_test.rb

require 'json'
require_relative 'finger_radius'

FIXTURE = File.expand_path('../fixtures/conformance.json', __dir__)
TOLERANCE = 1e-6

fixture = JSON.parse(File.read(FIXTURE))
failures = []
checks = 0

recognizer = FingerRadius::GestureRecognizer.new

fixture['gestures'].each do |item|
  landmarks = item['landmarks']
  expected = item['expected']
  actual = recognizer.classify(landmarks)
  checks += 1

  if actual[:gesture] != expected['gesture']
    failures << "#{item['id']}: gesture #{actual[:gesture]} != #{expected['gesture']}"
  end
  if actual[:extended_count] != expected['extended_count']
    failures << "#{item['id']}: extended_count #{actual[:extended_count]} != #{expected['extended_count']}"
  end
  expected['fingers'].each do |finger, want|
    got = actual[:fingers][finger]
    failures << "#{item['id']}: finger #{finger} #{got} != #{want}" if got != want
  end

  radii = FingerRadius.radii(landmarks)
  expected['radii'].each do |pair, want|
    got = radii[pair]
    checks += 1
    failures << "#{item['id']}: radius #{pair} #{got} != #{want}" if (got - want).abs > TOLERANCE
  end
end

kalman = fixture['kalman']
filter = FingerRadius::KalmanFilter1D.new(
  process_noise: kalman['process_noise'],
  measurement_noise: kalman['measurement_noise']
)
kalman['input'].each_with_index do |value, i|
  got = filter.update(value)
  want = kalman['expected'][i]
  checks += 1
  failures << "kalman[#{i}]: #{got} != #{want} (delta #{(got - want).abs})" if (got - want).abs > TOLERANCE
end

if failures.empty?
  puts "ruby: PASS - #{checks} checks against the Python fixture"
  exit 0
else
  puts "ruby: FAIL - #{failures.length} of #{checks} checks"
  failures.first(10).each { |f| puts "  #{f}" }
  exit 1
end
